libopenmpt: new port in audio

townba opened the pull request:
#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
macOS 13.1 22C65 arm64
Xcode 14.2 14C18

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. --> _(no existing tickets)_
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`? _(no existing tests)_
- [ ] tried a full install with `sudo port -vst install`? _(flag_ -t _does not work on my system for any port)_
- [x] tested basic functionality of all binary files?
- [x] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken? _(no variants)_

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

artkiver commented:
Seems to function OK here.

Admittedly, I am also running

macOS 13.1 22C65 arm64
Xcode 14.2 14C18

So, chances are that I wouldn't run into issues.

Regarding: 

` tried a full install with sudo port -vst install? (flag -t does not work on my system for any port)

At least in my experience, port -vst install only functions if I have previously installed the port.

e.g.

(omitting tail end of port -v install)```

x ./opt/local/bin/openmpt123
--->  Cleaning libopenmpt
--->  Removing work directory for libopenmpt
--->  Scanning binaries for linking errors
--->  No broken files found.
--->  No broken ports found.

`
```

Then running:
```

 port -vst install
--->  Computing dependencies for libopenmpt.
--->  Cleaning libopenmpt
--->  Removing work directory for libopenmpt
--->  Scanning binaries for linking errors
--->  No broken files found.
--->  No broken ports found.

```

Seems OK for me?

I would suggest (since I have been doing this to some other Portfiles I maintain) adding some livecheck bits perhaps?

e.g.


```
livecheck.type   regex
livecheck.url    https://lib.openmpt.org/libopenmpt/
livecheck.regex  "libopenmpt releases (\\d+\\.\\d+\\.\\d+),"
```

That seems to do the trick so that if you run:
```

 port -v livecheck
libopenmpt seems to be up to date
```

Meanwhile, for testing purposes I modified the Portfile version to 0.6.6 and would see this as an error:
```

 port -v livecheck
libopenmpt seems to have been updated (port version: 0.6.6, new version: 0.6.7)

```

As a means to verify that the livecheck parameters I set are functioning correctly.

You may also want to add a line such as:

`revision          0
`
Since it's been my experience sometimes that is useful?

Anyway, even without those suggestions, it looks good to me, but I don't have commit access!

townba commented:
Made the recommended changes and squashed the commits.

> At least in my experience, port -vst install only functions if I have previously installed the port.

On my system, it's a permissions/`sudo` issue. (I'm not too worried about it.) I get this error for all ports:

```
Command failed:  cd "/opt/local/var/macports/build/_opt_local_ports_audio_libopenmpt/libopenmpt/work" && /usr/bin/gzip -dc '/opt/local/var/macports/distfiles/libopenmpt/libopenmpt-0.6.7+release.autotools.tar.gz' | /usr/bin/tar -xf -
```

Adding `sudo` to the `tar` command (`… | sudo /usr/bin/tar -xf -`) gets the extraction to work.

> I would suggest (since I have been doing this to some other Portfiles I maintain) adding some livecheck bits perhaps?

Thank you! That's a nice addition.

> You may also want to add a line such as:
> 
> ```
> revision          0
> ```

Thank you again!

herbygillot commented on audio/libopenmpt/Portfile:
```diff
@@ -0,0 +1,34 @@
+version          0.6.7
+revision         0
+categories       audio devel multimedia
+platforms        darwin
```
`platforms darwin` is implicit, so this line can be dropped

herbygillot commented on audio/libopenmpt/Portfile:
```diff
@@ -0,0 +1,34 @@
+master_sites     https://lib.openmpt.org/files/libopenmpt/src
+distname         ${name}-${version}+release.autotools
+checksums        rmd160 edcaf6d9f0ea2cf3370423edd79afbdbbe2e0350 \
+    sha256 2174ac0f5a148ba684db768a47edf783eff9084fbca5fef6c997501643100163 \
```
Please align all the checksum lines if possible. Example:

```
checksums        rmd160  edcaf6d9f0ea2cf3370423edd79afbdbbe2e0350 \
                 sha256  2174ac0f5a148ba684db768a47edf783eff9084fbca5fef6c997501643100163 \
                 size    1556248
```

herbygillot commented on audio/libopenmpt/Portfile:
```diff
@@ -0,0 +1,34 @@
+livecheck.regex  "libopenmpt releases (\\d+\\.\\d+\\.\\d+),"
+
+depends_build    port:pkgconfig
+depends_lib      port:flac port:libogg port:libsndfile port:libvorbis \
```
For dependencies, it's ideal to have one per line:

```
depends_lib      port:flac \
                 port:libogg \
                 port:libsndfile \
                 port:libvorbis \
                 port:mpg123 \
                 port:portaudio \
                 port:zlib
```

sagamusix commented:
Hi, I'm one of the libopenmpt maintainers. Thanks for intending to add our library to macports!

Just a small note on the `livecheck.regex`. This regex is quite unstable and a quick glance at the libopenmpt website shows that it wouldn't have registered several updates in the previous year:
- If only a single libopenmpt branch gets a release (typically but not necessarily the most recent one), "libopenmpt *releases* x, y, z" would turn into "libopenmpt *release* x"
- Security updates are worded as "libopenmpt security updates x, y, z" or "libopenmpt security update x"

@manxorist might have more insight as he's responsible for the website, but I don't think that scanning the front page (or atom feed) is the best way to do this. I think the most stable way of extracting the most recent version of the latest stable branch would be to match the first `(\d+\.\d+\.\d+)` on https://lib.openmpt.org/libopenmpt/download/ instead. Maybe it's also possible to provide the latest version in a more machine-readable way, we'll have to see.

manxorist commented:
Yeah, using a regexp on the human-readable part of the website is probably not the most robust idea.

I am really not familiar with what is common practice for MacPorts/Livecheck (or even other distributions) in this area though.

Debian chooses to check full download directory listing (<https://lib.openmpt.org/files/libopenmpt/src/>) for href links, and parses out the newest version (see <https://salsa.debian.org/multimedia-team/libopenmpt/-/blob/debian/master/debian/watch>), which I would consider a more robust and future-proof approach than trying to parse the human-readable website.

I would suggest taking a similar approach (if possible).

Even though I have no intention in changing much of the human-readable website, updates to or switching of the static website generator that we use might some day introduce changes that might break things. The chances of server-generated http directory listings changing to a form that is not machine-parseable is probably less likely.

manxorist commented on audio/libopenmpt/Portfile:
```diff
@@ -0,0 +1,34 @@
+    --disable-debug \
+    --disable-dependency-tracking \
+    --disable-silent-rules \
+    --without-vorbisfile
```
`--without-vorbisfile` is highly likely wrong here. libopenmpt requires libogg, libvorbis, and libvorbisfile to properly load OggVorbis compressed samples in `OXM` and `MO3` formats.

libvorbisfile (and libvorbisenc) are normally provided by the same libvorbis package as libvorbis itself, so I think it should be available in MacPorts just fine (looking at <https://github.com/macports/macports-ports/blob/master/audio/libvorbis/Portfile>).

I am not using MacPorts myself, so I do not have a quick way to verify either situation right now.

I suggest you take a second look at that.
If you are seeing compilation errors with libvorbisfile, I will happily take a look and fix any issues upstream as soon as possible.

reneeotten commented:
@townba are you still interested in addressing the comments and trying to get this PR merged?

townba commented:
> @townba are you still interested in addressing the comments and trying to get this PR merged?

Yes. I had [a bit of a nasty surprise](https://blog.google/inside-google/message-ceo/january-update/) a few days after my last comment. I hope to get back to this soon.

townba commented:
PTAL

Made the following changes and squashed commits:
* Removed `--without-vorbisfile` option from `configure.args`.
* Fixed `homepage` copy-paste error.
* Modified `livecheck.url` to use recommended location. (Livecheck defaults work nicely with it.)
* Updated version to 0.7.0.
* Removed `platforms` line and reformatted `checksums` and `depends_lib` to match requested style.

reneeotten commented on audio/libopenmpt/Portfile:
```diff
@@ -0,0 +1,35 @@
+                 port:zlib
+
+configure.args \
+    --disable-debug \
```
from the CI log:
```
  configure: WARNING: unrecognized options: --disable-debug
```

reneeotten reviewed (commented):
Also from the CI log, I am a bit surprised to see the following:
```
/usr/bin/clang++ -std=gnu++11 -std=c++20 .......
```

I would assume you either set the standard to C++20 or C++11 but not both. Regardless, if any of these standards is required you should request a compatible compiler with:
```compiler.cxx_standard   2011```
If C++20 is needed some manual blacklisting of compilers will be necessary as the "2020" setting is not implemented yet (or more precise, not in a released version of `macports-base` yet)

manxorist commented:
> Also from the CI log, I am a bit surprised to see the following:
> 
> ```
> /usr/bin/clang++ -std=gnu++11 -std=c++20 .......
> ```

Well, great catch ... and what a nightmare of a rabbit hole!

Autoconf, in their infinite wisdom, decided that it would be a good idea to break existing behaviour and invented forcing some arbitrary outdated C++ standard version from 2011 over the toolchain default. In 2020, Autoconf version 2.70. Magnificent decision. 

See <https://git.savannah.gnu.org/gitweb/?p=autoconf.git;a=blob;f=NEWS;h=ba418d1af5da752de77a2c388f9af56f8f1bf6a4;hb=97fbc5c184acc6fa591ad094eae86917f03459fa#l259>:

> * Noteworthy changes in release 2.70 (2020-12-08) [stable]
[...]
> *** AC_PROG_CC now enables C2011 mode if the compiler supports it.
>
>   If not, it will fall back to C99 and C89, as before.  Similarly,
>   AC_PROG_CXX now enables C++2011 if available, falling back on C++98.

There appears to be ***no documentation*** on how to avoid this or work-around it, if one requires a newer C++ standard. See <https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.71/autoconf.html#C_002b_002b-Compiler>:

> If necessary, options are added to CXX to enable support for ISO Standard C++ features with extensions, preferring the newest edition of the C++ standard that is supported. Currently the newest edition Autoconf knows how to detect support for is ISO C++ 2011. After calling this macro, you can check whether the C++ compiler has been set to accept standard C++ by inspecting the shell variable ac_prog_cc_stdc. Its value will be ‘cxx11’ or ‘cxx98’, respectively, if the C++ compiler has been set to use the 2011 or 1990 edition of the C++ standard, and ‘no’ if the compiler does not support compiling standard C++ at all.

The established practice of choosing the desired C++ standard version with Autoconf has always been to use `AX_CXX_COMPILE_STDCXX` from the autoconf-archive package (<https://github.com/autoconf-archive/autoconf-archive/blob/master/m4/ax_cxx_compile_stdcxx.m4>), which is what libopenmpt does to check for C++20 support (or fall back to C++17 if C++20 is unavailable) (see <https://github.com/OpenMPT/openmpt/blob/libopenmpt-0.7.0/build/autotools/configure.ac#L26>).

So, if you do not want to see `-std=gnu++11 -std=c++20` in your Autotools build logs, you really need to complain to Autoconf (see <https://savannah.gnu.org/support/?110286>, <https://lists.gnu.org/archive/html/autoconf/2020-07/msg00010.html>, <https://lists.gnu.org/archive/html/autoconf/2020-07/msg00012.html>, and <https://savannah.gnu.org/support/?110394> for some roughly related discussions).

I submitted bug <https://savannah.gnu.org/support/index.php?110879> to Autoconf.

> I would assume you either set the standard to C++20 or C++11 but not both.

There is sadly absolutely nothing I can do in libopenmpt about it, as far as I can see.

> Regardless, if any of these standards is required

As for the required C++ standard version, libopenmpt (0.7) requires C++17 or C++20, and defaults to C++20.

reneeotten commented:
thanks for the comments @manxorist ! Perhaps the multiple specifications don't and if will only take the latest one. Not sure if it matters - I was just surprised about it. 

So we should request a compiler that at least supports 2017. @townba please add that, as well as removing the unrecognized configure argument.

manxorist commented:
> Perhaps the multiple specifications don't and if will only take the latest one. Not sure if it matters

On a quick look, I did not find it documented whether a later option value overrides a previous value of the same option for either GCC or Clang for `-std=`. In any case, this is rather common behavior in other cases for compilers though. And it is obviously the observed behavior also in this case.

It definitely matters whether C++11 or C++17/C++20 is chosen. However, the failure in this case should be trivially to detect as libopenmpt would completely fail to compile in C++11 mode. We would also notice such breakage in our own CI, so it should be kind-of safe for now. I am really struggling to even suggest any feasible workaround for this Autoconf weirdness at all here.

> I was just surprised about it.

So was I :)

reneeotten commented:
to be fair, I don't know much about this so perhaps it's "normal" 😉. Let's not worry about that for now, we should just add the correct compiler selection and we can merge it.

townba commented:
PTAL

Made the following changes and squashed commits:

* Added `compiler.cxx_standard 2017`.
* Removed `--disable-debug` option from `configure.args`.
